lib/Analysis/Utility.cpp

```diff
@@ -37,6 +37,55 @@ unsigned ReduceOpHelper::getThreadsReductionAxis() {
          triton::gpu::getWarpsPerCTA(srcLayout)[axis];
 }
 
+SmallVector<unsigned> ReduceOpHelper::getScratchConfigBasic() {
+  auto axis = op.axis();
+  auto smemShape = convertType<unsigned>(getSrcShape());
+  smemShape[axis] = std::min(smemShape[axis], getThreadsReductionAxis());
+  return smemShape;
+}
+
+SmallVector<SmallVector<unsigned>> ReduceOpHelper::getScratchConfigsFast() {
+  auto axis = op.axis();
+  SmallVector<SmallVector<unsigned>> smemShapes(3);
+
+  /// shared memory block0
+  smemShapes[0] = convertType<unsigned>(getSrcShape());
+  smemShapes[0][axis] = getInterWarpSize();
+
+  /// FIXME(Qingyi): This size is actually larger than required.
+  /// shared memory block1:
+  auto mod = op.getOperation()->getParentOfType<ModuleOp>();
+  unsigned numWarps = triton::gpu::TritonGPUDialect::getNumWarps(mod);
+  smemShapes[1].push_back(numWarps * 32);
+
+  /// FIXME(Qingyi): This requirement is actually not necessary, because it is
+  /// always smaller than smemShapes[0] shared memory block2
```

Comment:
Aha, this line was auto-reformatted by my clang-format 😅 It should have been:

```
  /// FIXME(Qingyi): This requirement is actually not necessary, because it is
  /// always smaller than smemShapes[0]
  /// shared memory block2
```

But it's okay to merge at the moment, since I will fix this function very soon.